include/ginkgo/extensions/kokkos/types.hpp

```diff
@@ -0,0 +1,288 @@
+
+        auto size = m.get_size();
+
+        return type<const ValueType>{
+            reinterpret_cast<const value_type_t<ValueType>*>(
+                m.get_const_values()),
+            Kokkos::LayoutStride{size[0], m.get_stride(), size[1], 1}};
+    }
+};
+
+
+template <typename ValueType, typename IndexType, typename MemorySpace>
+struct mapper<device_matrix_data<ValueType, IndexType>, MemorySpace> {
+    using index_mapper = mapper<array<IndexType>, MemorySpace>;
+    using value_mapper = mapper<array<ValueType>, MemorySpace>;
+
+    template <typename ValueType_c, typename IndexType_c>
+    struct type {
+        using index_array = typename index_mapper::template type<IndexType_c>;
+        using value_array = typename value_mapper::template type<ValueType_c>;
+
+        static type map(size_type size, IndexType_c* row_idxs,
+                        IndexType_c* col_idxs, ValueType_c* values)
+        {
+            return {index_mapper::map(row_idxs, size),
+                    index_mapper::map(col_idxs, size),
+                    value_mapper::map(values, size)};
+        }
+
+        index_array row_idxs;
+        index_array col_idxs;
+        value_array values;
+    };
+
+    static type<ValueType, IndexType> map(
+        device_matrix_data<ValueType, IndexType>& md)
+    {
+        assert_compatibility(md, MemorySpace{});
+        return type<ValueType, IndexType>::map(
+            md.get_num_stored_elements(), md.get_row_idxs(), md.get_col_idxs(),
+            md.get_values());
+    }
+
+    static type<const ValueType, const IndexType> map(
+        const device_matrix_data<ValueType, IndexType>& md)
+    {
+        assert_compatibility(md, MemorySpace{});
+        return type<const ValueType, const IndexType>::map(
+            md.get_num_stored_elements(), md.get_const_row_idxs(),
+            md.get_const_col_idxs(), md.get_const_values());
+    }
+};
+
+
+}  // namespace detail
+
+
+//!< specialization of gko::native for Kokkos
+template <typename MemorySpace>
+struct kokkos_type {
+    template <typename T>
+    static auto map(T* data)
+    {
+        return map(*data);
+    }
+
+    template <typename T>
+    static auto map(const std::unique_ptr<T>& data)
+    {
+        return map(*data);
+    }
+
+    template <typename T>
+    static auto map(const std::shared_ptr<T>& data)
+    {
+        return map(*data);
+    }
+
+    template <typename T>
+    static auto map(T&& data)
+    {
+        return detail::mapper<std::decay_t<T>, MemorySpace>::map(
+            std::forward<T>(data));
+    }
+};
+
+
+/**
+ * Maps Ginkgo object to a type compatible with Kokkos.
+ *
+ * @tparam T  The Ginkgo type.
+ * @tparam MemorySpace  The Kokkos memory space that will be used
+ *
+ * @param data  The Ginkgo object.
+ *
+ * @return  A wrapper for the Ginkgo object that is compatible with Kokkos
+ */
+template <typename T,
+          typename MemorySpace = Kokkos::DefaultExecutionSpace::memory_space>
```

Comment:
will Kokkos usually use memory space as variable or type?
I am thinking <MemorySpace, T> to using template rather than passing the variable. but it can not have the default template now

Reply:
Good point. From the [Kokkos docs](https://kokkos.org/kokkos-core-wiki/API/core/memory_spaces.html) there doesn't seem to be much useful stuff in a memory space class. Also the executor object have no function that returns a memory space object, so I will go with only the type parameter.